Bluetooth: audio: Merge ascs service and cb registration

fredrikdanebjer opened the pull request:
This change merges service registration and callback registration in ASCS. This should eliminate any weird middle states. Since ASCS can't properely function without callbacks being registered anyway, this shouldn't have any downsides.

Thalley commented on include/zephyr/bluetooth/audio/bap.h:
```diff
@@ -1291,62 +1275,54 @@ struct bt_bap_unicast_server_cb {
- * @return 0 in case of success or negative value in case of error.
- */
-int bt_bap_unicast_server_unregister_cb(const struct bt_bap_unicast_server_cb *cb);
+int bt_bap_unicast_server_unregister(const struct bt_bap_unicast_server_cb *cb);
```
This is kind of odd, but I understand why it's done like this. 

We could consider adding a `bt_bap_unicast_server` opaque struct similar to the broadcast sink and source, and return that as part of `bt_bap_unicast_server_register` that can then be used to unregister. 

It's weird to use the callback struct as the unregister parameter IMO

Thalley reviewed (changes requested):
Only reviews the header file for now

Thalley commented on tests/bluetooth/audio/ascs/src/main.c:
```diff
@@ -115,15 +116,15 @@ static void ascs_test_suite_after(void *f)
 	err = bt_bap_unicast_server_unregister();
-	zassert_equal(err, 0, "Unexpected err response %d", err);
+
+	/* In case of a failure due to pending state change, try again */
+	while (err != 0 && err != -EALREADY) {
+		/* Only -EBUSY indicate pending state change */
+		zassert_equal(err, -EBUSY, "unexpected err response %d", err);
+		k_sleep(K_MSEC(10));
+		err = bt_bap_unicast_server_unregister();
+	}
```
This isn't great IMO. 
`bt_bap_unicast_server_unregister` should be a synchronous action that should flush any pending state changes. 

Is that possible to do? IIRC there is some way to flush the system workqueue. Of course if this API is called in the system  workqueue thread we can't do that, can we?

fredrikdanebjer replied:
@Thalley 
>Is that possible to do? IIRC there is some way to flush the system workqueue. Of course if this API is called in the system workqueue thread we can't do that, can we?

Yes, you are correct: `Behavior is undefined if this function is invoked on work from a work queue running work.`

What do you suggest? You initial suggestion here is sound to me, to flush. I have a slightly harder time to see the strict need for registering the `bt_bap_unicast_server_unregister` in a work queue.

Thalley replied:
> I have a slightly harder time to see the strict need for registering the `bt_bap_unicast_server_unregister` in a work queue.

It's an API - We cannot control what thread it is being called in, as that is application specific. If it is an issue, we could in theory document that it shall not be called in the system workqueue thread and reject any calls to this functions from that thread, but that doesn't seem ideal either. 

Perhaps a better alternative is to modify ASCS so that it can be unregistered in a synchronous way, i.e. abort any work items with a sync and then don't start any new ones. I think today it will actually attempt to send a `released` notification during this called, but it doesn't make sense to send notifications for a service you are removing. We can also add the requirement, if it isn't already there, that this can only be done when all ASEs are in the idle state, which should ensure that we aren't triggering any state change notifications.

fredrikdanebjer replied:
> It's an API - We cannot control what thread it is being called in, as that is application specific. If it is an issue, we could in theory document that it shall not be called in the system workqueue thread and reject any calls to this functions from that thread, but that doesn't seem ideal either.

Yes I agree here that we cannot control this. My comment was rather about speculation of how likely it is that application will implement a registration from a work-queue. I'm nudging at that probably being unconventional and rare, adding the restriction and documentation that registration is not allowed in system work queue should be an ok tradeoff. 

>We can also add the requirement, if it isn't already there, that this can only be done when all ASEs are in the idle state, which should ensure that we aren't triggering any state change notifications.
We would then end up with a similar scenario on application side, i.e. that the application would have to error-check and retry until successful.

If you have a preference on what solution to use, then please let me know. Otherwise I would prefer the flush :)

Thalley replied:
> Yes I agree here that we cannot control this. My comment was rather about speculation of how likely it is that application will implement a registration from a work-queue. I'm nudging at that probably being unconventional and rare, adding the restriction and documentation that registration is not allowed in system work queue should be an ok tradeoff.

I don't think "This will rarely happen" is good enough if we want a mature stack. We need to work towards a mature and thread-safe stack, so we need to handle all corner cases that we are aware of. 

I think the best solution for this particular one is to ensure that `bt_bap_unicast_server_unregister` is a synchronous function that does not need to await for any state changes done by the system workqueue. 

This can be done by either rejecting it when all ASEs are not in the idle state (including no pending state changes), or by stopping all pending state changes and force-releasing all ASEs to move them to the IDLE state.

Thalley reviewed (commented):
Changes look good, but we do have a less-than-good thread issue with the unregister function. Any ideas?

fredrikdanebjer committed:
Bluetooth: audio: Merge ascs service and cb registration

This change merges service registration and callback registration in
ASCS. This should eliminate any weird middle states. Since ASCS can't
properely function without callbacks being registered anyway, this
shouldn't have any downsides.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

fredrikdanebjer commented:
I converted to draft - its been a while since it was updated. I will un-draft it when its ready for review.
